### DISCUSSION TITLE
unit and integration tests.  ffi=true and fii=false.

### DISCUSSION THREAD

**Discussion by YavorJJ** [Cyfrin/foundry-full-course-cu#1925] [General]
can someone take his time out to explain me the main diffrence between unit and integration tests and what are ffi=true and ffi=false. thanks for taking your time out :)

**Comment by EngrPips** [comment]
**UNIT TESTING**
This is the testing of a single unit in a component that makes up your whole system. An example of `unit testing` is when you test a single function in a specific smart contract to ensure that it does what it is supposed to do appropriately.

**INTEGRATION TESTING**
Integration testing is a type of [software testing] where software components are gradually integrated and then tested as a unified group. Usually, these components already work well individually, but they may break when integrated with other components. With integration testing, testers want to find defects that surface due to code conflicts between software modules when integrated. An example of Integration testing is when you write a test that tests a specific function in a contract, but in the process of testing that function, you use other components of your codebase such as the `HelperConfig,` `Deployment` script, and every other component that is needed to see how everything works together to achieve your test purpose.

**ffi=true**
When you set `ffi` to true in foundry, you have basically given foundry the permission to execute any arbitrary command on your device, which can be very risky, so it is recommended that it is always set to `false` except if you need to set it to `true.`

**Reply by YavorJJ** [reply]
Appreciate it!

**Reply by EngrPips** [reply]
Anytime, Friend.
